Upgrade to new SDK

austinkelleher opened the pull request:
Sorry for the big PR. This ended up basically being a rewrite of the previous integration. A lot of refactors, additional docs, introduced `google_domain` as well because we already were leveraging the domain data.

aiwilliams commented on src/constants.ts:
```diff
@@ -0,0 +1,79 @@
+import {
```
Love this file as a "registry" of all thing things this generates. 🎉

aiwilliams replied:
Adopted this in https://github.com/JupiterOne/graph-artifactory/pull/4 👍

austinkelleher commented:
A few other changes/notes:

- In the previous integration, we were passing the service account key file contents in through the `invocationArgs`. That doesn't exist in the new SDK. I updated this integration to take a `serviceAccountKeyFile` as integration config. I propose that we tweak the managed integration to push our managed service account key file into the integration config when it's being invoked. This makes development on `graph-google` consistent with other integrations (e.g. `graph-google-cloud`) as well.

- Removed user custom attributes (see here: https://github.com/JupiterOne/graph-google/blob/master/src/converters/UserEntityConverter.ts#L179). The way those are currently being stored on the entity does not match data-model requirements. Here is an example of a custom attribute: `J1_Groups_-_j1dev_-_jupiterone-prod-us`.

- Buildings were previously being duplicated for every user. (e.g. RTP entity created for every user who is in RTP office). I changed this to dedup and only create the building once.

- Introduce `google_domain` while still supporting existing flow of domains on the `google_account` entity. Example useful query:

```
find Domain with verified=false
find google_domain with primary=true
```

- I didn't add recording tests for most resources because I didn't think we wanted to expose internal user data. Consider using a separate shared gsuite account for integration dev?

- Experimented with top-level `constants.ts` file with entity and relationship metadata

- Probably can share some testing utilities between google cloud and gsuite integrations. Ended up copying some of the testing code for now.

- Discovered that there is a "resources" API that we may want to consider hitting down the road. You can store "buildings" and other resources in there. Worth mentioning. 

- Currently the data-model requires that `_key` must be at least 10 characters. I think we should consider softening that limit. For example, in the gsuite case, I would have liked to have made the `google_domain` `_key` value just the domain name. There are domain names that are less than 10 characters, so I just prefixed it with the `_type` for now (we decided this is an anti-pattern, but I didn't have much of a choice).

- Currently now `HOSTS` in the data-model. The old integration has the following relationship: `google_site_hosts_user`. Should we add `HOSTS` to the data-model or was that an oversight? In the meantime, I just changed to `HAS`. I want to come to a consensus on that before this PR merges.

- The old integration referenced `config.googleAccountName`, which looks to not even be used in the UI or referenced in the deployed project. I removed it. It's worth noting that this brought up an interesting question - should we allow _optional_ config values in the integration config? Currently, the execution process will throw if you have a `instanceConfigFields` value that isn't supplied to the invocation. I think we should consider adding a `required: false` option?

aiwilliams commented:
Regarding your notes, do you think we can focus on those as separate issues/discussions, or would you like them discussed in this PR?

austinkelleher commented:
@aiwilliams Main question that we should address before merging is the second to last bullet:

```
Currently now HOSTS in the data-model. The old integration has the following relationship: google_site_hosts_user. Should we add HOSTS to the data-model or was that an oversight? In the meantime, I just changed to HAS. I want to come to a consensus on that before this PR merges.
```

aiwilliams commented:
> Removed user custom attributes (see here: https://github.com/JupiterOne/graph-google/blob/master/src/converters/UserEntityConverter.ts#L179). The way those are currently being stored on the entity does not match data-model requirements. Here is an example of a custom attribute: J1_Groups_-_j1dev_-_jupiterone-prod-us.

I think this is a great change. I doubt there was much value in property names like that.

> I didn't add recording tests for most resources because I didn't think we wanted to expose internal user data. Consider using a separate shared gsuite account for integration dev?

I think it would be good to have a separate GSuite account for integration development. 👍 

> Currently the data-model requires that _key must be at least 10 characters. I think we should consider softening that limit. For example, in the gsuite case, I would have liked to have made the google_domain _key value just the domain name. There are domain names that are less than 10 characters, so I just prefixed it with the _type for now (we decided this is an anti-pattern, but I didn't have much of a choice).

I think we should relax that as well. It was introduced as a sanity check, but with all the great work done in the integration SDK otherwise, such as ensuring uniqueness within the `scope` of the `_integrationInstanceId`, it doesn't really matter if the `_key` is one character, as long as it's unique within the `scope`.

> Currently now HOSTS in the data-model. The old integration has the following relationship: google_site_hosts_user. Should we add HOSTS to the data-model or was that an oversight? In the meantime, I just changed to HAS. I want to come to a consensus on that before this PR merges.

Is that a word in the GSuite docs/UI? I think your change is good unless there is some argument for `HOSTS`.

> The old integration referenced config.googleAccountName, which looks to not even be used in the UI or referenced in the deployed project. I removed it. It's worth noting that this brought up an interesting question - should we allow optional config values in the integration config? Currently, the execution process will throw if you have a instanceConfigFields value that isn't supplied to the invocation. I think we should consider adding a required: false option?

Hmm, good question. The Feroot integration fields declare [a `boolean` property](https://github.com/JupiterOne/graph-feroot/blob/3cd010704effffc82fd64a8df7c9ccffe9a5b102/src/instanceConfigFields.ts#L8) that I made optional in the UI config. I don't know off hand whether the UI will ensure there is a `false` put in there or not 😳  Seems we should support an optional. Also, would love to make the UI driven from these fields instead of maintaining a similar structure in the UI code.

ndowmon commented:
@austinkelleher what docstringn would you add to the `data-model` to describe `HOSTS`? e.g.
```
/**
 * A relationship between a group or service Entity and a member of that Entity.
 * Closely related to or interchangeable with CONTAINS.
 */
HAS = 'HAS'
```

Maybe answering that question can help to determine if there is value in adding `HOSTS`. Otherwise, I think `HAS` is a fine alternative.

ndowmon commented:
Unbelievable amount of work in this PR, thanks for taking it on! Motivosity deserving 💸 

For the record, and I know this is hard to do sometimes, but it's a challenging PR to review since it's all in one commit.

erkangz commented:
HAS looks good to me instead of HOSTS. We have Host as an entity. Let's avoid that as a relationship.

j1-internal-automation commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`v3.16.4`](https://github.com/JupiterOne/graph-google/releases/tag/v3.16.4) :rocket:
